### DISCUSSION TITLE
How to check the mass_balance when BiGG metabolites have multiple formulas

### DISCUSSION THREAD

**Discussion by dtmvandenberg** [opencobra/cobrapy#1045] [Q&A] [answered]
Dear Moritz and Matthias,

Do you know how to ensure that reaction.check_mass_balance() does not give false positives when the metabolites come from the BiGG database and have multiple formulas, like f.i. phosphate (pi) has as formula HPO4;HO4P.
I have made a model using CarveMe and checked the metabolites and mass balance of reactions as such:
![Screenshot from 2021-02-19 15-41-59](https://user-images.githubusercontent.com/62939535/108518808-48c82000-72c9-11eb-9347-1ed9cf1bfe05.png)

Kind regards,

Dennis

**Comment by Midnighter** [comment, marked as answer]
Hmm, I had actually not noticed this before and it's certainly not SBML compliant. This will also present a problem for [memote.io](https://memote.io) which just looks at the formula directly. The only thing I can immediately think of, is to pre-process metabolite formulas and pick one of them. You could also get fancy and let OpenBabel/RDKit/ChemAxon give you a formula at specific pH but that seems like overkill.

**Comment by cdiener** [comment]
I think this is a bug in CarveME. @cdanielmachado is that intentional?

**Reply by matthiaskoenig** [reply]
This is reported in https://github.com/cdanielmachado/carveme/issues/101

**Reply by dtmvandenberg** [reply]
Thanks for the info, I will try to write a script to do this automatically then and post it at the CarveMe issue. Maybe @cdanielmachado can then use it to curate his universe files.

**Reply by cdanielmachado** [reply]
I initially decided not to exclude any information obtained from BiGG, including the multiple formulae and charges per metabolite. In retrospective, this was a bad decision. I am currently working on a new release that gets the formula and charge from metanetx instead. Not sure if it is more reliable than BiGG but at least the values are unique.

**Reply by dtmvandenberg** [reply]
Ok than I will just curate them manually in my model  instead of trying to do it generally for all metabolites as that will probably become a spaghetti of if statements ;)

**Reply by Midnighter** [reply]
@cdanielmachado one of the major inputs for MetaNetX compounds is ChEBI. So I hope the information is good but I have no solid info on it. As part of the work for equilibrator, I've written a pipeline that parses the tables from MetaNetX and puts the information in a database structure (see https://github.com/opencobra/cobra-component-models/ for the models). Maybe that can jump start your work?

**Reply by cdanielmachado** [reply]
Thanks, I will take a look.
